ls1intum/Artemis
# `General`: Add type filtering to average course scores chart for instructors

### Checklist
#### General


#### Client

### Motivation and Context


This PR changes some existing functionality and adds some details to the average score chart on course statistics page
### Description


- **Data label**: Above the chart bars, a data label now displays now the represented average exercise score numerically.
- **Content order**: The exercises are now ordered ascending in their average score.
- **Dynamic chart coloring**: Previously, all bars representing an average score >= 90% were colored green and all bars representing an average score <= 25% were colored red. We decided to change this logic. Now, the in average lowest performing third of the course exercises is colored red and the in average best performing third is colored green. Note, that the exercises might not be necessarily separated in exact thirds. This depends on the general amount of exercises and the average score distribution. If for example 5 out of 10 exercises share the lowest score, it would not make much sense to declare 3 of them as lowest third and 2 of them as part of the average third. 
- **Color legend**: A color legend is added in order to explain the color code. By clicking on an arbitrary legend entry, filtering of the chart is possible:
     - When selecting an interval that is already selected, all other selected intervals get deselected
     - When selecting an interval that is currently the only one selected, all intervals get selected again
     - When selecting an interval that is currently deselected, it is added to the current filter
- **Type filter**: Additionally, it is now possible to filter the chart by exercise types.

Additional change:

- I changed the page structure a bit and moved the tab switcher manipulating the displayed time span below the average scores chart and separated it with a light grey border from the other charts. I think this reduces the confusion of the user since the current layout might imply that the average scores chart is time span sensitive as well and can be manipulated by the tab switcher (which is not the case apparently).

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with some contributions

1. Log in to Artemis
2. Navigate to Course Administration
3. Select "Statistics" of the corresponding course and inspect the first chart
4. Make sure that the exercises are ordered by their average score
5. Play around with the filter options (Type and interval filter). Make sure both work as expected. Note, that the interval filter will always reset the type filter. This is done in order to prevent undesired side effects. 

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





| File name | Branch coverage (%) | Line coverage (%) |
| -- | -- | -- |
| `exercise-scores-chart.component.ts` | 79.41 | 98 |
| `chart-exercise-type-filter.directive.ts` | 50 | 100 |
| `statistics-average-scores-graph.component.ts` | 79.41 | 95.86 |
### Screenshots


![image](https://user-images.githubusercontent.com/80622272/154841722-64b436a0-f035-49db-b057-c1abe621f271.png)
